src/lib/Lower.hs

```diff
@@ -426,22 +412,37 @@ vectorizeOp op = do
       x' <- if vx /= v then ensureVarying sx else return x
       y' <- if vy /= v then ensureVarying sy else return y
       VVal v <$> emitOp (BinOp opk x' y')
-    CastOp (VVal Uniform ty) (VVal vx x) -> do
+    MiscOp (CastOp (VVal Uniform ty) (VVal vx x)) -> do
       ty' <- case vx of
         Uniform    -> return ty
         Varying    -> getVectorType ty
         Contiguous -> return ty
         ProdStability _ -> error "Unexpected cast of product type"
-      VVal vx <$> emitOp (CastOp ty' x)
-    PtrOffset (VVal Uniform ptr) (VVal Contiguous off) -> do
-      VVal Contiguous <$> emitOp (PtrOffset ptr off)
-    PtrLoad (VVal Contiguous ptr) -> do
+      VVal vx <$> emitOp (MiscOp $ CastOp ty' x)
+    MemOp (PtrOffset (VVal Uniform ptr) (VVal Contiguous off)) -> do
+      VVal Contiguous <$> emitOp (MemOp $ PtrOffset ptr off)
+    MemOp (PtrLoad (VVal Contiguous ptr)) -> do
       BaseTy (PtrType (addrSpace, a)) <- getType ptr
       BaseTy av <- getVectorType $ BaseTy a
-      ptr' <- emitOp $ CastOp (BaseTy $ PtrType (addrSpace, av)) ptr
-      VVal Varying <$> emitOp (PtrLoad ptr')
+      ptr' <- emitOp $ MiscOp $ CastOp (BaseTy $ PtrType (addrSpace, av)) ptr
+      VVal Varying <$> emitOp (MemOp $ PtrLoad ptr')
     _ -> error $ "Can't vectorize op: " ++ pprint op'
 
+    -- IndexRef (VVal Uniform ref) (VVal Contiguous i) -> do
+    --   TC (RefType _ (TabTy tb a)) <- getType ref
+    --   vty <- getVectorType $ case hoist tb a of
+    --     HoistSuccess a' -> a'
+    --     HoistFailure _  -> error "Can't vectorize dependent table application"
+    --   VVal Contiguous <$> emitOp (VectorSubref ref i vty)
+    -- Place (VVal vref ref) sval@(VVal vval val) -> do
+    --   VVal Uniform <$> case (vref, vval) of
+    --     (Uniform   , Uniform   ) -> emitOp $ Place ref val
+    --     (Uniform   , _         ) -> error "Write conflict? This should never happen!"
+    --     (Varying   , _         ) -> error "Vector scatter not implemented"
+    --     (Contiguous, Varying   ) -> emitOp $ Place ref val
+    --     (Contiguous, Contiguous) -> emitOp . Place ref =<< ensureVarying sval
+    --     _ -> error "Not implemented yet"
+
```

Comment:
Why commented out?

Reply:
Good catch. It's a refactoring artifact, but I didn't notice it because vectorize is dead (until Norman's patch revives it).